util/automation.py

```diff
@@ -24,6 +39,139 @@
 DATA_DIR  = os.path.join(BaseDirectory.xdg_data_home, 'deepspeech_wer')
 CKPT_DIR  = os.path.join(DATA_DIR, 'checkpoint')
 
+class GPUUsage(Thread):
+    def __init__(self, csvfile=None):
+        super(GPUUsage, self).__init__()
+
+        self._cmd        = [ 'nvidia-smi', 'dmon', '-d', '1', '-s', 'pucvmet' ]
+        self._names      = []
+        self._units      = []
+        self._stop       = False
+        self._process    = None
+
+        self._csv_output = csvfile or os.environ.get('ds_gpu_usage_csv', self.make_basename(prefix='ds-gpu-usage', extension='csv'))
+
+    def make_basename(self, prefix, extension):
+        # Let us assume that this code is executed in the current git clone
+        return '%s.%s.%s.%s' % (prefix, git.Repo('.').git.describe('--always', '--dirty', '--abbrev'), int(time()), extension)
+
+    def stop(self):
+        print "Ending nvidia-smi monitoring: PID", self._process.pid
+        self._process.terminate()
+        print "Ended nvidia-smi monitoring ..."
+
+    def run(self):
+        print "Starting nvidia-smi monitoring"
+
+        # If the system has no CUDA setup, then this will fail.
+        try:
+            self._process = subprocess.Popen(self._cmd, stdout=subprocess.PIPE)
+        except OSError as ex:
+            print "Unable to start monitoring, check your environment:", ex
+            return
+
+        writer = None
+        with open(self._csv_output, 'w') as f:
+            for line in iter(self._process.stdout.readline, ''):
```

Comment:
Just a question: How does this loop behave when `self._process.terminate()` is called?